### PR TITLE
Llm adblock

### PR DESCRIPTION
**Pull request by posix4e**
I haven't carefully reviewed yet, but it looks like we an amazing start.

Ported from https://github.com/posix4e/puppet/pull/45.

TODO

- [x] Basic python end to end test acceptance test showing it working
- [x] Basic java unit tests for the internal logic
- [x] Basic Java end to end test acceptance test showing it working
- [x] allow gradio to use gpt4all
- [x] Get rid of any direct calls to openai, and switch everything over to gpt4all

### CODE REVIEW

**Comment by posix4e** [comment]
Ported from https://github.com/posix4e/puppet/pull/45

**Comment by urbit-pilled** [comment]
looks like saucelabs is still using the old apk file in the E2E tests and it looks like the E2E tests gets triggered twice (push and pull_request)

**Comment by vkolgi** [comment]
@urbit-pilled that's right. I am working on a PR which will help test the android app at PR level, before merging to master.
This should be done by EOD. 

Are these tests running in your local with android app ? There is a flag in tests to run it against your local appium server.

Since I do not want you to be blocked on your work, we can temporarily upload this PR specific APK to saucelabs and make tests run against it. That however means uploading it manually every time there is a change in android code.

**Comment by urbit-pilled** [comment]
@vkolgi Yea, I saw the flag to run appium locally, and I've been using that to write the selenium tests.
I'm fine with the temporary solution of running tests against the current apk build, I don't think the apk is going to be changed much after the latest commit (for this PR) and I'll request an apk re-upload if I make any breaking changes.

**Comment by vkolgi** [comment]
@urbit-pilled It's now handled in another PR. Will wait for @posix4e to have a look at the PR. 
So, once that's approved, you should be able to rebase and have the tests running against PR specific apk.

**Comment by vkolgi** [comment]
@urbit-pilled the PR has been merged now. Can you bring in those changes to your branch?

File: puppet/app/src/main/res/drawable/ic_pause_black_24dp.xml
Comment (urbit-pilled) [inline puppet/app/src/main/res/drawable/ic_pause_black_24dp.xml:1]: dont need this anymore

File: tests/test_sauce_labs.py
Comment (urbit-pilled) [inline tests/test_sauce_labs.py:1]: using https://up20-puppet-staging.hf.space as a staging url

File: puppet/app/src/main/res/drawable/ic_pause_black_24dp.xml
Comment (urbit-pilled) [reply puppet/app/src/main/res/drawable/ic_pause_black_24dp.xml:1]: removed

**Comment by urbit-pilled** [comment]
@posix4e PR is ready to be merged

**Comment by vkolgi** [comment]
@urbit-pilled I am testing this out locally. Just give me a day to finalize.

**Comment by vkolgi** [comment]
Also the app while running test is crashing. Attaching the video here. @posix4e 
As soon as the test pressed the button Turn Off VPN, app crashed. Attached crash log below:
At timestamp 0.31sec the app crashes and android home screen appears.

```
java.lang.InterruptedException
	at java.lang.Thread.sleep(Native Method)
	at java.lang.Thread.sleep(Thread.java:442)
	at java.lang.Thread.sleep(Thread.java:358)
	at com.ttt246.puppet.AdblockVPNThread.run(AdblockVPNThread.kt:103)
	at java.lang.Thread.run(Thread.java:923)
```

https://github.com/posix4e/puppet/assets/2462783/ecc891ce-4eb7-46a1-aaad-3fa45b172d8b

**Comment by urbit-pilled** [comment]
I think that's just the end of the unit test, the test presses the start vpn button, then the stop vpn button and then it automatically closes the program after the unit test is done.

**Comment by vkolgi** [comment]
@urbit-pilled at the end of the test, the app is crashing. So it looks like test is done but the app crashes as soon as disable VPN is pressed. Attaching the screenshot from the saucelabs here.

Screenshot 2023-09-02 at 10 05 59 PM

**Comment by urbit-pilled** [comment]
You are right @vkolgi , I should have handled the `InterruptedException` better. Since I don't have saucelabs can you check to see if the latest commit fixes this?
```kotlin
} catch (e: InterruptedException) {
            Log.i(TAG, "Vpn Thread interrupted")
            //throw e <--- deleted this and instead reinterrupted the thread
            Thread.currentThread().interrupt();
```
https://github.com/posix4e/puppet/commit/c4de696586998cd0cd4a8282bec5168ad3eff519#diff-b0aa8eda0216e995473db3013d9cbd3e36c5c831e737ae601afcb42c5f4c3619R108-R109

**Pull request closed without merging**